File: scripts/ParseData.m
function data = ParseData(filename)
    % File size
    s = dir(filename);         
    filesize = s.bytes;

    % Define the number of variables per record
    bytesPerRecord = 6 * 8 + 8 + 8 + 8 + 8; % 6 doubles (8 bytes each) + 1 uint64 (8 bytes) + 2 double + 1 uint64
    
    % Compute the total number of records
    totalRecords = floor(filesize / bytesPerRecord);
    
    % Memory-map the file
    m = memmapfile(filename, 'Format', 'uint8', 'Writable', false);
    
    % Extract raw bytes as a single vector
    rawData = m.Data(1:totalRecords * bytesPerRecord);
    % Preallocate structured output
    data = struct();
    data.time = zeros(1, totalRecords);
    data.cpu = zeros(1, totalRecords);
    data.ram = zeros(1, totalRecords);
    data.disk = zeros(1, totalRecords);
    data.bandwidth = zeros(1, totalRecords);
    data.fpga = zeros(1, totalRecords);
    data.turnedOnMachineCount = zeros(1, totalRecords, 'uint64');
    data.averagePowerConsumption = zeros(1, totalRecords);
    data.totalPowerConsumption = zeros(1, totalRecords);
    data.numberOfSLAVs = zeros(1, totalRecords, 'uint64');

    % **Extract interleaved data correctly using strided access**
    for i = 1:totalRecords
        startIdx = (i-1) * bytesPerRecord + 1;
        
        data.time(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.cpu(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.ram(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.disk(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.bandwidth(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.fpga(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.turnedOnMachineCount(i) = typecast(rawData(startIdx : startIdx + 7), 'uint64');
        startIdx = startIdx + 8;
        
        data.averagePowerConsumption(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.totalPowerConsumption(i) = typecast(rawData(startIdx : startIdx + 7), 'double');
        startIdx = startIdx + 8;
        
        data.numberOfSLAVs(i) = typecast(rawData(startIdx : startIdx + 7), 'uint64');
    end
end

File: scripts/ParseData.py
import numpy as np
import matplotlib.pyplot as plt

def parse_binary(filename):
    # Define the structure: 10 doubles per record (each 8 bytes)
    dtype = np.dtype([
        ('time', 'f8'), ('cpu', 'f8'), ('ram', 'f8'), ('disk', 'f8'),
        ('bandwidth', 'f8'), ('fpga', 'f8'), ('turnedOnMachineCount', 'u8'),
        ('averagePowerConsumption', 'f8'), ('totalPowerConsumption', 'f8'), ('numberOfSLAVs', 'u8')
    ])
    
    # Read binary data efficiently
    data = np.fromfile(filename, dtype=dtype)
    
    return data

def plot_data(data):
    fig, ax = plt.subplots(3, 2, figsize=(12, 8))

    ax[0, 0].plot(data['time'], data['cpu'], label="CPU")
    ax[0, 0].plot(data['time'], data['ram'], label="RAM")
    ax[0, 0].plot(data['time'], data['disk'], label="Disk")
    ax[0, 0].plot(data['time'], data['bandwidth'], label="Bandwidth")
    ax[0, 0].plot(data['time'], data['fpga'], label="FPGA")
    ax[0, 0].set_title("Resource Utilization")
    ax[0, 0].legend()
    
    ax[1, 0].plot(data['time'], data['turnedOnMachineCount'], label="Machines")
    ax[1, 0].set_title("Turned On Machines")
    
    ax[1, 1].plot(data['time'], data['numberOfSLAVs'], label="SLA Violations")
    ax[1, 1].set_title("SLA Violations")
    
    ax[2, 0].plot(data['time'], data['averagePowerConsumption'], label="Avg Power Consumption")
    ax[2, 0].set_title("Average Power Consumption")
    
    ax[2, 1].plot(data['time'], data['totalPowerConsumption'] / 1e3, label="Total Power (kW)")
    ax[2, 1].set_title("Total Power Consumption")
    
    for i in range(3):
        for j in range(2):
            ax[i, j].grid()
            ax[i, j].legend()
    
    plt.tight_layout()
    plt.show()

# Load data
data = parse_binary("/Users/oddogan/MSc/CDC/CloudDataCenter_Simulator/runs/ILPDQN128_10e3.bin")

plot_data(data);